# Networks: structure, evolution & processes
**Internet Analytics - Lab 2**

---

**Group:** *N*

**Names:**

* *Sandra Djambazovska(224638)*
* *Anh Nghia Khau(223613)*

---

#### Instructions

*This is a template for part 2 of the lab. Clearly write your answers, comments and interpretations in Markodown cells. Don't forget that you can add $\LaTeX$ equations in these cells. Feel free to add or remove any cell.*

*Please properly comment your code. Code readability will be considered for grading. To avoid long cells of codes in the notebook, you can also embed long python functions and classes in a separate module. Don’t forget to hand in your module if that is the case. In multiple exercises, you are required to come up with your own method to solve various problems. Be creative and clearly motivate and explain your methods. Creativity and clarity will be considered for grading.*

---

## 2.2 Network sampling

#### Exercise 2.7: Random walk on the Facebook network

In [1]:
import requests
import numpy as np
URL_TEMPLATE = 'http://iccluster118.iccluster.epfl.ch:5050/v1.0/facebook?user={user_id}'
s = 'f30ff3966f16ed62f5165a229a19b319'
N = 10000

In [2]:
def naive_solution(T, s):
    next_ = s
    i = 0
    acc = 0
    
    while (i < T):
        if (i % 1000 == 0 and i != 0):
            print('Naive solution for avg age at iter {it} is {a}'.format(it = i, a = float(acc)/i))
        data = requests.get((URL_TEMPLATE.format(user_id=next_))).json()
        friends = data['friends']
        acc += data['age']
        next_ = friends[np.random.randint(len(friends))]
        i += 1
    return float(acc)/i

In [ ]:
print (naive_solution(N, s))

The estimated age is XXX after 10000 iterations.

#### Exercise 2.8

In [26]:
def randomwalk_solution(T, s):
    numerator = 0.0
    denumerator = 0.0
    i = 0

    while (i < T):
        if (i % 1000 == 0 and i != 0):
            print('Random wark solution for avg age at iter {it} is {a}'.format(it = i, a = numerator/denumerator))
        data = requests.get((URL_TEMPLATE.format(user_id=s))).json()
        friends = data['friends']
        outDe =  len(friends)
        numerator += float(data['age'])/outDe
        denumerator += 1.0/outDe              
        i +=1        
        s = friends[np.random.randint(len(friends))]
    return numerator/denumerator


In [27]:
print (randomwalk_solution(N, s))

Random wark solution for avg age at iter 1000 is 42.58026429417872
Random wark solution for avg age at iter 2000 is 43.41714851850858
Random wark solution for avg age at iter 3000 is 43.33376543321675
Random wark solution for avg age at iter 4000 is 43.53909998704922
Random wark solution for avg age at iter 5000 is 44.341070250513745
Random wark solution for avg age at iter 6000 is 44.374082984518004
Random wark solution for avg age at iter 7000 is 44.14930253475083
Random wark solution for avg age at iter 8000 is 43.85392283378637
Random wark solution for avg age at iter 9000 is 43.80545715935731
43.79057838854075


1. Our estimation is far, far from the true average age.

2. Because we always start from the same node, whose age is 19 so most of him/her's friends are the same age and the friends of their friends also. 

3. Not knowing the total number of nodes (n) and edges (m) we had to use the estimator function without knowledge of n,m(seen in class)

 ## F =  $ \frac{\sum_{t} f(X_t) / d_{X_t}}{\sum_{t} 1 / d_{X_t}} $
 
 With the new strategy we have around 43.5 which is really close to the true average.
